src/runtime/dynlink/src/context/load.rs

```diff
@@ -0,0 +1,393 @@
+
+use crate::{
+    compartment::{Compartment, CompartmentId},
+    context::engine::{LoadDirective, LoadFlags},
+    library::{BackingData, CtorInfo, Library, LibraryId, UnloadedLibrary},
+    tls::TlsModule,
+    DynlinkError, DynlinkErrorKind, HeaderError,
+};
+
+use super::{engine::ContextEngine, Context, LoadedOrUnloaded};
+
+impl<Engine: ContextEngine> Context<Engine> {
+    // Collect information about constructors.
+    pub(crate) fn get_ctor_info(
+        &self,
+        libname: &str,
+        elf: &elf::ElfBytes<'_, NativeEndian>,
+        base_addr: usize,
+    ) -> Result<CtorInfo, DynlinkError> {
+        let dynamic = elf
+            .dynamic()?
+            .ok_or_else(|| DynlinkErrorKind::MissingSection {
+                name: "dynamic".to_string(),
+            })?;
+
+        // If this isn't present, just call it 0, since if there's an init_array, this entry must be present in valid ELF files.
+        let init_array_len = dynamic
+            .iter()
+            .find_map(|d| {
+                if d.d_tag == DT_INIT_ARRAYSZ {
+                    Some((d.d_val() as usize) / size_of::<usize>())
+                } else {
+                    None
+                }
+            })
+            .unwrap_or_default();
+
+        // Init array is a pointer to an array of function pointers.
+        let init_array = dynamic.iter().find_map(|d| {
+            if d.d_tag == DT_INIT_ARRAY && d.clone().d_ptr() != 0 {
+                Some(base_addr + d.d_ptr() as usize)
+            } else {
+                None
+            }
+        });
+
+        // Legacy _init call. Supported for, well, legacy.
+        let leg_init = dynamic.iter().find_map(|d| {
+            if d.d_tag == DT_INIT && d.clone().d_ptr() != 0 {
+                Some(base_addr + d.d_ptr() as usize)
+            } else {
+                None
+            }
+        });
+
+        if dynamic.iter().any(|d| d.d_tag == DT_PREINIT_ARRAY)
+            && dynamic
+                .iter()
+                .find(|d| d.d_tag == DT_PREINIT_ARRAYSZ)
+                .is_some_and(|d| d.d_val() > 0)
+        {
+            warn!("{}: PREINIT_ARRAY is unsupported", libname);
+        }
+
+        debug!(
+            "{}: ctor info: init_array: {:?} len={}, legacy: {:?}",
+            libname, init_array, init_array_len, leg_init
+        );
+        Ok(CtorInfo {
+            legacy_init: leg_init.unwrap_or_default(),
+            init_array: init_array.unwrap_or_default(),
+            init_array_len,
+        })
+    }
+
+    // Load (map) a single library into memory via creating two objects, one for text, and one for data.
+    fn load<N>(
+        &mut self,
+        comp_id: CompartmentId,
+        unlib: UnloadedLibrary,
+        idx: NodeIndex,
+        n: N,
+    ) -> Result<Library<Engine::Backing>, DynlinkError>
+    where
+        N: FnMut(&str) -> Option<Engine::Backing>,
+    {
+        let backing = self.engine.load_object(&unlib, n)?;
```

Comment:
Can you describe `n` a little better and in other places? (load.rs, load_object, etc.)